### DISCUSSION TITLE
생성자 참조를 위한 함수형 인터페이스 생성

### DISCUSSION THREAD

**Discussion by coalong** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#47] [CHAPTER 3 람다 표현식 - Q&A] [answered]
p.120

퀴즈 3-7.

### Q. 
- 지금까지 인수가 없거나, 하나 또는 둘인 생성자를 생성자 참조로 바꾸는 방법을 살펴봤다.
- Color(int, int, int) 처럼 인수가 세 개인 생성자의 참조를 사용하려면 어떻게 해야 할까?

### A.
> 생성자 참조와 일치하는 시그니처를 갖는 함수형 인터페이스가 필요하다. 현재 이런 시그니처를 갖는 함수형 인터페이스는 제공되지 않으므로 우리가 직접 다음과 같은 함수형 인터페이스를 만들어야 한다.

``` java
// 사용자 정의 함수형 인터페이스
public interface TriFunction<T, U, V, R> {
    R apply(T t, U u, V v);
}

// 새로운 생성자 참조를 사용
TriFunction<Integer, Integer, Integer, Color> colorFactory = Color::new;

```
---

**Comment by Irisation23** [comment, marked as answer]
질문하신 의도를 파악해 보면은 TriFunction으로 새로운 함수형 인터페이스를 구현하는 것 이외에 
자바에서 제공하는 `Funciton Interface` 만으로 구현할 수 없는지에 대한 질문의도로 파악하고 답변을 드려보겠습니다.

먼저 `Color` 클래스는 아래와 같은 구조로 되어있습니다.

```java
package com.hello.modern.chapter3.domain;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@ToString
public class Color {

    private final int red;
    private final int green;
    private final int blue;

    public Color(int[] colors) {
        this.red = colors[0];
        this.green = colors[1];
        this.blue =  colors[2];
    }
}

```

`Function`을 통해 `Color`클래스를 생성하는 방법은 아래와 같습니다.

```java
Function<int[], Color> colorFactory = Color::new;
Color color = colorFactory.apply(new int[]{255, 0, 0}); // red
System.out.println(color);
```

<img width="463" alt="image" src="https://github.com/Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action/assets/53285909/983fa89f-0b35-4939-8e18-61bfca48464c">

하지만 과연 Color의 생성자를 새로운 정의를 하기 위해 `int[]`을 사용하는 것이 괜찮은 로직일까요?

따라서 이 상황에서는 제안해주셨던 `TriFunction`을 이용한다면 아래의 이점을 볼 수 있습니다.

1. 생성자에 하드코딩 된 배열의 index를 참조를 하지 않아도 된다. (e.g - colors[0], colors[1] ...)
2. 함수형 인터페이스가 추구하는 정의 중 하나인 간결성을 보장할 수 있다.
3. 생성자에 단순 참조성 배열을 정의하지 않아도 된다. (e.g - `@RequiredArgsConstructor` 에 알맞는 생성자만 사용)

```java
package com.hello.modern.chapter3.domain;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@ToString
public class Color {

    private final int red;
    private final int green;
    private final int blue;
}
```

```java
package com.hello.modern.chapter3;

import com.hello.modern.chapter3.domain.Color;

public class Main {

    public static void main(String[] args) {
        TriFunction<Integer, Integer, Integer, Color> colorFactory = Color::new;
        Color color = colorFactory.apply(255, 255, 255);

        System.out.println("color = " + color);
    }

    public interface TriFunction<T, U, V, R> {

        R apply(T t, U u, V v);
    }
}

```

해당 상황에 생성자 메서드 참조를 통한 정의를 위함이라면 `TriFunction` 으로 정의 하는것도 괜찮아 보입니다.

하지만 단순히 생성자 만을 정의하기 위한 함수형 인터페이스의 정의라면, 오히려 코드 가독성을 해칠 수 있다는 생각을 해보며 이만 답변을 마무리 하겠습니다.
